### PR TITLE
os/driver/audio: bringup syu645b audio driver

### PR DESCRIPTION
**Pull request by ritesh55555**
Add driver support for syu645b audio amplifier

### CODE REVIEW

File: os/drivers/audio/syu645b.c [hunk 1]
```diff
@@ -0,0 +1,1055 @@
...
+        FAR struct i2c_dev_s *dev = priv->i2c;
+        FAR struct i2c_config_s *syu645b_i2c_config = &(priv->lower->i2c_config);
+
+        reg[0] = regaddr_h;
+        reg[1] = regaddr_l;
+        reg[2] = regval;
+
+        ret = i2c_write(dev, syu645b_i2c_config, (uint8_t *)reg, 3);
+        if (ret < 0) {
+			auddbg("Error, cannot write reg 0x%x/0x%x/0x%x\n", regaddr_h, regaddr_l, regval);
```
Comment (edwakuwaku) [inline os/drivers/audio/syu645b.c:102]: alignment seems wrong~

File: os/drivers/audio/syu645b.c [hunk 2]
```diff
@@ -0,0 +1,1055 @@
...
+ *   Get the audio device capabilities
+ *
+ ****************************************************************************/
+static int syu645b_getcaps(FAR struct audio_lowerhalf_s *dev, int type, FAR struct audio_caps_s *caps)
+{
+#if !defined(CONFIG_AUDIO_EXCLUDE_VOLUME) || !defined(CONFIG_AUDIO_EXCLUDE_TONE)
+	FAR struct syu645b_dev_s *priv = (FAR struct syu645b_dev_s *)dev;
+#endif
+	/* Validate the structure */
+
```
Comment (edwakuwaku) [inline os/drivers/audio/syu645b.c:270]: maybe removing unnecessary change new line

File: os/drivers/audio/syu645b.c [hunk 3]
```diff
@@ -0,0 +1,1055 @@
...
+ *
+ ****************************************************************************/
+FAR struct audio_lowerhalf_s *syu645b_initialize(FAR struct i2c_dev_s *i2c, FAR struct i2s_dev_s *i2s, FAR struct syu645b_lower_s *lower)
+{
+
+	FAR struct syu645b_dev_s *priv;
+	/* Sanity check */
+	DEBUGASSERT(i2c && lower && i2s);
+
+	auddbg("I2s dev addr is 0x%x\n", i2s);
```
Comment (edwakuwaku) [inline os/drivers/audio/syu645b.c:1019]: audvdbg?

File: os/board/rtl8730e/src/rtl8730e_syu645b.c [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+#include <tinyara/audio/i2s.h>
+#include <tinyara/audio/syu645b.h>
+#include <tinyara/gpio.h>
+#include "amebasmart_i2s.h"
+
+#include "objects.h"
+#include "PinNames.h"
+#include "gpio_api.h"
+
+extern FAR struct i2s_dev_s *amebasmart_i2s_initialize(uint16_t port, bool is_reinit);
```
Comment (Taejun-Kwon) [inline os/board/rtl8730e/src/rtl8730e_syu645b.c:40]: actually i2s also should apply interface like i2c_initialize.
can you add this after you finish this?

File: os/board/rtl8730e/src/rtl8730e_syu645b.c [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+			auddbg("ERROR: Failed to initialize the SYU645B\n");
+			ret = -ENODEV;
+			goto errout_with_syu645b;
+		}
+		auddbg("calling init lower done\n");
+		/* Finally, we can register the PCM/SYU645B/I2C/I2S audio device.
+		 */
+		auddbg("registering audio device here\n");
+		ret = audio_register(devname, syu645b);
+		auddbg("registrred the device\n");
```
Comment (Taejun-Kwon) [inline os/board/rtl8730e/src/rtl8730e_syu645b.c:208]: typo

File: os/board/rtl8730e/src/rtl8730e_syu645b.c [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+ *
+ * Input Parameters:
+ *   minor - The input device minor number
+ *
+ * Returned Value:
+ *   Zero is returned on success.  Otherwise, a negated errno value is
+ *   returned to indicate the nature of the failure.
+ *
+ ****************************************************************************/
+int rtl8730e_syu645b_initialize(int minor)
```
Comment (Taejun-Kwon) [inline os/board/rtl8730e/src/rtl8730e_syu645b.c:135]: please update rtl8730e_boot.c too.

File: os/drivers/audio/syu645b.c [hunk 7]
```diff
@@ -0,0 +1,1055 @@
...
+
+	if (!priv->running) {
+		/* Add the new buffer to the tail of pending audio buffers */
+		syu645b_takesem(&priv->devsem);
+		sq_addlast((sq_entry_t *)&apb->dq_entry, &priv->pendq);
+		syu645b_givesem(&priv->devsem);
+		return OK;
+	}
+	
+	ret = I2S_SEND(priv->i2s, apb, syu645b_txcallback, priv, SYU645B_I2S_TIMEOUT_MS);
```
Comment (Taejun-Kwon) [inline os/drivers/audio/syu645b.c:749]: Last parameter, actually it should be calculated based on samplerate & bitrate & buffer size.
For example, if sample rate is 48k and bitreate is 16bit, and buffer size is 8192, then 12 times of i2s should be executed within 1s.
In this case timeout value should be set as 80~90 ms.

File: os/drivers/audio/syu645b.c [hunk 8]
```diff
@@ -0,0 +1,1055 @@
...
+
+		break;
+
+	/* Provide capabilities of our INPUT & OUTPUT unit */
+	case AUDIO_TYPE_OUTPUT:
+		caps->ac_channels = SYU645B_CHANNELS;
+		switch (caps->ac_subtype) {
+		case AUDIO_TYPE_QUERY:
+			/* Report the Sample rates we support */
+			caps->ac_controls.b[0] = AUDIO_SAMP_RATE_TYPE_48K;
```
Comment (Taejun-Kwon) [inline os/drivers/audio/syu645b.c:313]: it support more sample rate.
please check.

**Review by Taejun-Kwon — COMMENTED** [review]
Thanks for your effor, I will check more when I have some time.

File: os/drivers/audio/Kconfig [hunk 9]
```diff
@@ -180,6 +180,15 @@ config AUDIO_ALC1019
                 NOTE: This driver also depends on both I2C and I2S support although
                 that dependency is not explicit here.
 
+config AUDIO_SYU645B
+        bool "SYU645B DAC audio chip"
+        default n
+        depends on AUDIO && I2S && I2C && SCHED_WORKQUEUE
+        ---help---
+                Select to enable support for the SYU645B DAC Audio codec
+                NOTE: This driver also depends on both I2C and I2S support although
```
Comment (kishore-sn) [inline os/drivers/audio/Kconfig:189]: Remove this note pls. Dependency is indeed clearly visible.

File: os/drivers/audio/syu645b.c [hunk 10]
```diff
@@ -0,0 +1,1055 @@
...
+        .pause = syu645b_pause,                 /* pause          */
+        .resume = syu645b_resume,               /* resume         */
+#endif
+        .allocbuffer = NULL,                    /* allocbuffer    */
+        .freebuffer = NULL,                     /* freebuffer     */
+        .enqueuebuffer = syu645b_enqueuebuffer, /* enqueue_buffer */
+        .cancelbuffer = syu645b_cancelbuffer,   /* cancel_buffer  */
+        .ioctl = syu645b_ioctl,                 /* ioctl          */
+        .read = NULL,                           /* read           */
+        .read = NULL,                           /* write          */
```
Comment (kishore-sn) [inline os/drivers/audio/syu645b.c:84]: Read api initialised twice instead of write.

File: os/board/rtl8730e/src/Make.defs [hunk 11]
```diff
@@ -56,6 +56,10 @@ ifeq ($(CONFIG_AUDIO_ALC1019),y)
 CSRCS += rtl8730e_alc1019.c
 endif
 
+ifeq ($(CONFIG_AUDIO_SYU645B),y)
```
Comment (Taejun-Kwon) [inline os/board/rtl8730e/src/Make.defs:59]: please add this config to 8730/audio defconfig
And below logic too.
```
#ifdef CONFIG_AUDIO_ALC1019
	if (rtl8730e_alc1019_initialize(0) != 0) {
		lldbg("ALC1019 initialization failed\n");
	}
#endif
#ifdef CONFIG_AUDIO_SYU645B
	if (rtl8730e_syu645b_initialize(0) != 0) {
		lldbg("syu645b initialization failed\n");		
	}
#endif
```

File: os/board/rtl8730e/src/rtl8730e_syu645b.c [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+		g_syu645binfo.lower.control_hw_reset(true); /*power on*/
+		up_mdelay(10);
+		g_syu645binfo.lower.control_hw_reset(false); /*does nothing*/
+		up_mdelay(10);
+
+		/* Now we can use these I2C and I2S interfaces to initialize the
+		* SYU645B which will return an audio interface.
+		*/
+		/* Create a device name */
+		snprintf(devname, sizeof(devname), "pcmC%uD%u%c", minor, 0, 'p');
```
[the remark this answers is not in the document]
Comment (Taejun-Kwon) [reply os/board/rtl8730e/src/rtl8730e_syu645b.c:193]: Currnent Path is right, you should update description

File: os/drivers/audio/syu645b.c [hunk 13]
```diff
@@ -0,0 +1,1132 @@
...
+
+        ret = i2c_write(dev, syu645b_i2c_config, (uint8_t *)reg, 5);
+        if (ret < 0) {
+		auddbg("Error, cannot write reg 0x%x/0x%x\n", regaddr, regval);
+        } 
+
+        return ret;
+}
+
+static int syu645b_writereg_3byte(FAR struct syu645b_dev_s *priv, uint8_t regaddr, uint8_t regval[4])
```
Comment (Taejun-Kwon) [inline os/drivers/audio/syu645b.c:146]: duplicated.

File: os/drivers/audio/syu645b.c [hunk 14]
```diff
@@ -0,0 +1,1132 @@
...
+ ************************************************************************************/
+
+static int syu645b_exec_i2c_script(FAR struct syu645b_dev_s *priv, t_codec_init_script_entry *script, uint32_t size)
+{
+	uint32_t i;
+        uint16_t ret = 0;
+
+         for (i = 0; i < size; i++) {
+
+		if (script[i].type == SYU645B_REG_D_2BYTE) {
```
Comment (Taejun-Kwon) [inline os/drivers/audio/syu645b.c:201]: we can reduce line of code here.
```
for (i = 0; i < size; i++) {
	reg[0] = script[i].addr;
	if (script[i].type == SYU645B_REG_D_2BYTE) {
		reg[1] = script[i].val[0];
	} else if (script[i].type == SYU645B_REG_D_3BYTE) {
		reg[1] = script[i].val[0];
		reg[2] = script[i].val[1];
	} else if (script[i].type == SYU645B_REG_D_5BYTE) {
		...
	}
	...

	ret = i2c_write(dev, syu645b_i2c_config, (uint8_t *)reg, script[i].type);
	if (script[i].delay > 0) {
		up_mdelay(script[i].delay);
	}
	...
}

File: os/drivers/audio/syu645b.c [hunk 15]
```diff
@@ -0,0 +1,1132 @@
...
+
+        if (priv->running) {
+                audvdbg("volume=%u mute=%u\n", priv->volume, priv->mute);
+                if (priv->mute) {
+			syu645b_exec_i2c_script(priv, codec_init_mute_on_script, sizeof(codec_init_mute_on_script) / sizeof(t_codec_init_script_entry));
+                } else {
+			if (priv->volume == 0) {
+				codec_set_master_volume_script[0].val[0] = 0;
+			} else {
+				codec_set_master_volume_script[0].val[0] = SYU645B_HW_VOL_MAX - (SYU645B_SPK_VOL_MAX - priv->volume) * SYU645B_HW_VOL_STEP;
```
Comment (Taejun-Kwon) [inline os/drivers/audio/syu645b.c:265]: please check volume logic in datasheet more. Media framework has 10 of volume level, volume level should be adjust properly.

File: os/drivers/audio/syu645b.c [hunk 16]
```diff
@@ -0,0 +1,1132 @@
...
+        reg[1] = regval;
+
+        ret = i2c_write(dev, syu645b_i2c_config, (uint8_t *)reg, 2);
+        if (ret < 0) {
+		auddbg("Error, cannot write reg 0x%x/0x%x\n", regaddr, regval);
+        }
+        return ret;
+}
+
+static int syu645b_writereg_3byte(FAR struct syu645b_dev_s *priv, uint8_t regaddr, uint8_t regval[4])
```
Comment (abhishek-samsung) [inline os/drivers/audio/syu645b.c:106]: all of this looks redundant, I think better way is to create a function like syu645b_writereg_nbyte and pass the length as argument.
```
static int syu645b_writereg_nbyte(FAR struct syu645b_dev_s *priv, uint8_t regaddr, uint8_t* regval, size_t nbytes);
```

File: os/drivers/audio/syu645b.c [hunk 17]
```diff
@@ -0,0 +1,1132 @@
...
+		priv->nchannels = caps->ac_channels;
+		priv->bpsamp = caps->ac_controls.b[2];
+
+		audvdbg("    Number of channels: 0x%x\n", priv->nchannels);
+		audvdbg("    Sample rate:        0x%x\n", priv->samprate);
+		audvdbg("    Sample width:       0x%x\n", priv->bpsamp);
+
+		/* Reconfigure the FLL to support the resulting number or channels,
+		 * bits per sample, and bitrate.
+		 */
```
Comment (abhishek-samsung) [inline os/drivers/audio/syu645b.c:536]: here, we will have to add changes to set i2s and device to the required sampling rate and data width.

File: os/board/rtl8730e/src/rtl8730e_syu645b.c [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+/* i2c config */
+#if CONFIG_RTL8730E_BOARD_REVISION >= 5
+#define SYU645B_I2C_PORT		2
+#else
+#define SYU645B_I2C_PORT		1
+#endif
```
Comment (abhishek-samsung) [inline os/board/rtl8730e/src/rtl8730e_syu645b.c:48]: This doesnt make sense here as this amp is added from rev6 onwards.

File: os/drivers/audio/syu645b.c [hunk 19]
```diff
@@ -0,0 +1,1043 @@
...
+                } else if (script[i].type == SYU645B_REG_D_3BYTE) {
+                        reg[1] = script[i].val[0];
+                        reg[2] = script[i].val[1];
+                } else if (script[i].type == SYU645B_REG_D_5BYTE) {
+                        reg[1] = script[i].val[0];
+                        reg[2] = script[i].val[1];
+                        reg[3] = script[i].val[2];
+                        reg[4] = script[i].val[3];
+                } else {
+                        auddbg("Error, script type is not supported\n")
```
Comment (neel-samsung) [inline os/drivers/audio/syu645b.c:118]: Please check semi-colon at the end

File: os/drivers/audio/syu645b.c [hunk 20]
```diff
@@ -0,0 +1,1043 @@
...
+        /* if no audio device object return */
+        if (!priv) {
+                auddbg("Error, Device's private data Not available\n");
+                return;
+        }
+
+        if (priv->running) {
+                audvdbg("volume=%u mute=%u\n", priv->volume, priv->mute);
+                if (priv->mute) {
+			syu645b_exec_i2c_script(priv, codec_init_mute_on_script, sizeof(codec_init_mute_on_script) / sizeof(t_codec_init_script_entry));
```
Comment (neel-samsung) [inline os/drivers/audio/syu645b.c:171]: where do you declare ```codec_init_mute_on_script```? Please check.

Comment (ritesh55555) [reply os/drivers/audio/syu645b.c:171]: done, thanks for the check

**Review by kishore-sn — DISMISSED** [review]
Needs work. Please check pending comments.

File: os/drivers/audio/syu645b.c (hunk 7 above)
Comment (jsdosa) [reply os/drivers/audio/syu645b.c:739]: @ritesh55555 did you check it?

File: os/drivers/audio/syu645b.c (hunk 8 above)
Comment (jsdosa) [reply os/drivers/audio/syu645b.c:313]: @ritesh55555

File: os/drivers/audio/syu645b.c (hunk 7 above)
Comment (ritesh55555) [reply os/drivers/audio/syu645b.c:739]: updated

File: os/drivers/audio/syu645b.c (hunk 8 above)
Comment (ritesh55555) [reply os/drivers/audio/syu645b.c:313]: currently it is using 48K

File: os/board/rtl8730e/src/rtl8730e_syu645b.c (hunk 4 above)
Comment (ritesh55555) [reply os/board/rtl8730e/src/rtl8730e_syu645b.c]: as it is bsp code,i can discuss with rtk and add this

File: os/board/rtl8730e/src/rtl8730e_syu645b.c [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+
+#include "objects.h"
+#include "PinNames.h"
+#include "gpio_api.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+/* i2c config */
+#if CONFIG_RTL8730E_BOARD_REVISION >= 5
```
Comment (jsdosa) [inline os/board/rtl8730e/src/rtl8730e_syu645b.c:44]: is it right?
>= 6

Comment (ritesh55555) [reply os/board/rtl8730e/src/rtl8730e_syu645b.c:44]: yes, port should be 2 for 6

File: os/drivers/audio/syu645b.c (hunk 8 above)
Comment (Taejun-Kwon) [reply os/drivers/audio/syu645b.c:313]: Let's expand later.

File: os/drivers/audio/syu645b.c (hunk 7 above)
Comment (Taejun-Kwon) [reply os/drivers/audio/syu645b.c:739]: I mean it should be caluclated dynamically, not fixed value.

File: os/drivers/audio/syu645b.c (hunk 15 above)
Comment (Taejun-Kwon) [reply os/drivers/audio/syu645b.c:176]: What I meaning is below.
volume 0 : mute
volume 1 : 10% of max volume
volume 2: 20% of max volume

So we need extra algorithm.

**Review by r-prabu — APPROVED** [review]

**Review by Taejun-Kwon — APPROVED** [review]

**Comment by Taejun-Kwon** [comment]
Please raise PR related to my comment.
